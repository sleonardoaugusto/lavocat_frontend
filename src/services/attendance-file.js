import useModal from '@/hooks/useModal'
import UploadFile from '@/services/providers/uploadFile'

const modal = useModal()

class AttendanceFile {
  constructor(http) {
    this.http = http
  }
  async getFilesByAttendanceId(attendanceId) {
    if (attendanceId) {
      return await this.http
        .get(`/attendances/${attendanceId}/attendance-files/`)
        .then(resp => resp.data)
        .catch(() =>
          modal.open({
            component: 'SnackBar',
            props: { type: 'error', text: 'Erro ao carregar arquivos.' }
          })
        )
    }
  }
  async uploadAttendanceFiles(attendanceId, files) {
    if (files && files.length) {
      const filesParsed = []

      for (let i = 0; i < files.length; i++) {
        if (files[i] instanceof File) {
          const fd = new FormData()
          fd.append('attendance', attendanceId)
          fd.append('file', files[i])
          fd.append('filename', files[i].name)
          filesParsed.push(fd)
        }
      }

      const uploadService = new UploadFile(this.http)
      await uploadService.uploadFiles(filesParsed, 'post', '/attendance-files/')
    }
  }
  async updateAttendanceFile(attendanceFileId, filename) {
    const data = { filename: filename }
    return await this.http
      .patch(`/attendance-files/${attendanceFileId}/`, data)
      .then(resp => {
        modal.open({
          component: 'SnackBar',
          props: { type: 'success', text: 'Arquivo renomeado.' }
        })
        return resp.data
      })
      .catch(() =>
        modal.open({
          component: 'SnackBar',
          props: { type: 'error', text: 'Erro ao renomear arquivo.' }
        })
      )
  }
  async deleteAttendanceFile(attendanceFileId) {
    return await this.http
      .delete(`/attendance-files/${attendanceFileId}/`)
      .then(() => {
        modal.open({
          component: 'SnackBar',
          props: { type: 'success', text: 'Arquivo deletado.' }
        })
      })
      .catch(e => {
        console.error(`Ocorreu um erro ${e}`)
        modal.open({
          component: 'SnackBar',
          props: { type: 'error', text: `Ocorreu um erro ${e}` }
        })
      })
  }
}

const factory = httpClient => new AttendanceFile(httpClient)

export default factory
