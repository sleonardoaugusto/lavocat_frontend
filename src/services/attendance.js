export default httpClient => ({
  create: async(data) => {
    async function uploadFiles(files, attendanceId) {
      async function upload(formData) {
        const requestData = {
          method: 'post',
          url: '/attendance-files/',
          data: formData,
          headers: {
            'Content-type': 'multipart/form-data'
          }
        }
        const response = await httpClient.request(requestData)
        return response.data
      }

      const promises = files.map(async f => {
        const fd = new FormData()
        fd.append('attendance', attendanceId)
        fd.append('file', f)
        await upload(fd)
      })
      await Promise.all(promises)
    }

    const files = data.files
    delete data.files

    const response = await httpClient.post('/attendances/', data)
    const { id } = response.data

    await uploadFiles(files, id)

    return {
      data: response.data
    }
  }
})
