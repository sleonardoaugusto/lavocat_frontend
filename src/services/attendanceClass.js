import useModal from '@/hooks/useModal'
import Upload from '@/services/upload'

const modal = useModal

class Attendance {
  constructor(http) {
    this.http = http
  }
  createAttendance(data) {
    return this.http.post(data).then(resp => {
      const { id } = resp.data
      const { files } = data
      this.uploadAttendanceFiles(id, files)

      modal.open({
        component: 'SnackBar',
        props: { type: 'success', text: 'Atendimento criado!' }
      })
      return resp.data
    })
  }
  uploadAttendanceFiles(attendanceId, files) {
    if (files && files.length) {
      const filesParsed = files.map(f => {
        const fd = new FormData()
        fd.append('attendance', 1)
        fd.append('file', f)
        return fd
      })

      const uploadService = new Upload(this.http)
      uploadService.uploadFiles(filesParsed, 'post', '/attendance-files/')
    }
  }
}

const factory = httpClient => new Attendance(httpClient)

export default factory
