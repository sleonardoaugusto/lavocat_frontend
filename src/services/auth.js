import useModal from '@/hooks/useModal'

const modal = useModal()

class Auth {
  constructor(http) {
    this.http = http
  }
  async login(data) {
    return await this.http
      .post('/api/token/', data)
      .then(resp => {
        const { access, is_superuser } = resp.data
        return { access, is_superuser }
      })
      .catch(() => {
        modal.open({
          component: 'SnackBar',
          props: { type: 'error', text: 'Credenciais invÃ¡lidas.' },
        })
        return { access: null, is_superuser: null }
      })
  }
}

const factory = httpClient => new Auth(httpClient)

export default factory
