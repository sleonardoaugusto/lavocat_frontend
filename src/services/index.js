import axios from 'axios'
import AttendanceService from './attendance'
import useModal from '@/hooks/useModal'

const BASE_URL = process.env.VUE_APP_SERVICE_URL
const httpClient = axios.create({
  baseURL: BASE_URL
})

httpClient.interceptors.response.use(function(response) {
  return response
}, function(error) {
  const modal = useModal()
  modal.open({ component: 'SnackBar', props: { type: 'error', text: 'Erro no servidor' } })
  return Promise.reject(error)
})

export default {
  auth: AttendanceService(httpClient)
}
