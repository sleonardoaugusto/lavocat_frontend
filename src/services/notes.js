import useModal from '@/hooks/useModal'

const modal = useModal()

class Notes {
  constructor(http) {
    this.http = http
  }
  async getNotes(attendanceId) {
    return await this.http
      .get(`/attendances/${attendanceId}/notes/`)
      .then(resp => resp.data)
  }
  async updateNote(attendanceId, noteId, data) {
    return await this.http
      .patch(`/attendances/${attendanceId}/notes/${noteId}/`, data)
      .then(resp => {
        modal.open({
          component: 'SnackBar',
          props: { type: 'success', text: 'Notas salvas com sucesso!' },
        })
        return resp.data
      })
      .catch(err => {
        modal.open({
          component: 'SnackBar',
          props: { type: 'error', text: 'Erro ao salvar notas' },
        })
        throw new Error(err)
      })
  }
}

const factory = httpClient => new Notes(httpClient)

export default factory
