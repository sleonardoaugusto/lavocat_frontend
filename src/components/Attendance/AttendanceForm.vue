<template>
  <v-form>
    <v-overlay ref="overlay" :value="busy">
      <v-progress-circular indeterminate size="64" />
    </v-overlay>
    <v-row>
      <v-col cols="12" md="6">
        <v-text-field
          id="customer-name"
          ref="customerName"
          label="Nome do Cliente"
          v-model="$v.form.customer_name.$model"
          :error-messages="errorMessage('customer_name')"
        />
      </v-col>
      <v-col cols="12" md="6">
        <v-text-field
          id="document-id"
          ref="documentId"
          label="CPF"
          type="tel"
          v-mask="'###.###.###-##'"
          v-model="$v.form.document_id.$model"
          :error-messages="errorMessage('document_id')"
        />
      </v-col>
      <v-col cols="12" md="6">
        <ServicesOptions
          ref="servicesOptions"
          @changed="options => (form.services_provided = options)"
          :selected="form.services_provided"
        />
      </v-col>
      <v-col cols="12" md="6">
        <AttendanceFiles
          ref="attachments"
          @changed="files => (form.files = files)"
          :value="form.files"
        />
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="12">
        <v-textarea
          id="resume"
          ref="resume"
          label="RelatÃ³rio"
          v-model="form.resume"
          rows="10"
        />
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="12">
        <v-textarea
          id="status-resume"
          ref="statusResume"
          label="Resumo de Status"
          v-model="form.status_resume"
          rows="10"
        />
      </v-col>
    </v-row>
    <v-row>
      <v-col class="d-flex justify-space-between">
        <v-btn ref="previousBtn" id="previousBtn" href="/atendimentos">
          Voltar
        </v-btn>
        <v-btn
          ref="submitBtn"
          id="submit"
          @click="onSubmit"
          :loading="busy"
          :color="colorStyle"
          class="white--text"
          depressed
          right
        >
          {{ submitBtnLabel }}
        </v-btn>
      </v-col>
    </v-row>
  </v-form>
</template>

<script>
import { required } from 'vuelidate/lib/validators'
import validations from '@/mixins/formValidations'
import services from '@/services'
import { cpfValidator } from '@/utils/validators'
import { clearDocumentId, objToSelect } from '@/utils/formatters'
import AttendanceFiles from '@/components/Attendance/AttendanceFiles'
import ServicesOptions from '@/components/Attendance/ServicesOptions'

export default {
  name: 'AttendanceForm',
  components: { ServicesOptions, AttendanceFiles },
  mixins: [validations],
  props: {
    value: {
      type: Object,
      default: () => {},
    },
    update: {
      type: Boolean,
      default: false,
    },
    busy: {
      type: Boolean,
      default: false,
    },
  },
  data: () => ({
    form: {
      customer_name: null,
      document_id: null,
      files: [],
      resume: null,
      status_resume: null,
      services_provided: [],
    },
  }),
  validations: {
    form: {
      customer_name: { required },
      document_id: { cpfValidator },
    },
  },
  async created() {
    if (!this.update) {
      const cache = JSON.parse(localStorage.getItem('atttendance-form-cache'))
      if (cache && Object.keys(cache)) this.form = { ...cache }
    }
  },
  methods: {
    onSubmit() {
      this.touch()
      if (this.formIsReady) {
        const data = this.parseForm()
        this.$emit('submit', data)
      }
    },
    parseForm() {
      return {
        ...this.form,
        document_id: clearDocumentId(this.form.document_id),
      }
    },
  },
  computed: {
    submitBtnLabel() {
      return this.update ? 'Salvar' : 'Cadastrar'
    },
    colorStyle() {
      return this.update ? 'primary' : 'green'
    },
  },
  watch: {
    value: {
      deep: true,
      handler(val) {
        if (this.update) this.form = { ...this.form, ...val }
      },
    },
    form: {
      deep: true,
      handler(val) {
        if (!this.update) {
          const cache = { ...val }
          delete cache['files']
          localStorage.setItem('atttendance-form-cache', JSON.stringify(cache))
        }
      },
    },
  },
}
</script>

<style scoped></style>
