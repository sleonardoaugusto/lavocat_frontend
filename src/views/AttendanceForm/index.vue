<template>
  <v-form>
    <BaseHeading text="Atendimento"/>
    <v-row>
      <v-col cols="12" md="6" sm="12">
        <v-text-field
          id="customer-name"
          ref="customerName"
          label="Nome"
          v-model="$v.form.customer_name.$model"
          :error-messages="errorMessage('customer_name')"
        />
      </v-col>
      <v-col cols="12" md="6" sm="12">
        <v-text-field
          id="document-id"
          ref="document-id"
          label="CPF"
          v-model.number="$v.form.document_id.$model"
          :error-messages="errorMessage('document_id')"
        />
      </v-col>
      <v-col cols="12" md="6" sm="12">
        <v-file-input
          id="files"
          ref="files"
          label="Anexos"
          v-model="form.files"
          multiple
          chips
        />
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="12">
        <v-textarea
          label="RelatÃ³rio"
          value=""
          rows="10"
        />
      </v-col>
    </v-row>
    <v-row>
      <v-col class="text-right">
        <v-btn @click="submit" id="submit" color="success" right>
          Cadastrar
        </v-btn>
      </v-col>
    </v-row>
  </v-form>
</template>

<script>
import BaseHeading from '@/components/base/BaseHeading'
import { required } from 'vuelidate/lib/validators'
import validations from '@/mixins/validations'
import services from '@/services'

export default {
  name: 'AttendanceForm',
  components: {
    BaseHeading
  },
  mixins: [validations],
  data: () => ({
    form: {
      customer_name: null,
      document_id: null,
      files: []
    }
  }),
  validations: {
    form: {
      customer_name: { required },
      document_id: { required }
    }
  },
  methods: {
    async submit() {
      this.touch()
      if (this.formIsReady) {
        await services.auth.create(this.form)
      }
    }
  }
}
</script>

<style scoped>

</style>
