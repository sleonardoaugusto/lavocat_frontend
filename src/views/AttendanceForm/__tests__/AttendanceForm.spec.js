import Vue from 'vue'
import Vuetify from 'vuetify'
import Vuelidate from 'vuelidate'
import AttendanceForm from '@/views/AttendanceForm'
import { mount } from '@vue/test-utils'
import faker from 'faker'
import services from '@/services/.'
import axios from 'axios'

axios.post = jest.fn()

describe('<AttendanceForm />', () => {
  Vue.use(Vuetify)
  Vue.use(Vuelidate)
  let wrapper
  let vuetify

  beforeEach(() => {
    vuetify = new Vuetify()
    wrapper = factory()
  })

  const factory = opts => mount(AttendanceForm, { Vue, vuetify, ...opts })

  test('Must render component correctly', async() => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Page must render with one file input element', () => {
    const fileInput = '.v-file-input__text'
    expect(wrapper.findAll(fileInput)).toHaveLength(1)
  })

  test.each([['customerName', null], ['document-id', null]])('%s field must be valid', async(field, msg) => {
    expect(wrapper.findComponent({ ref: field }).vm.errorMessages).toBe(
      msg
    )
  })

  test.each([['customerName', 'Campo obrigatório'], ['document-id', 'Campo obrigatório']])('%s field must be invalid', async(field, msg) => {
    await wrapper.find('#submit').trigger('click')
    expect(wrapper.findComponent({ ref: field }).vm.errorMessages).toBe(
      msg
    )
  })

  test('Must send form data', async() => {
    const spy = jest.spyOn(services.auth, 'create')
    const data = await fillForm()
    await wrapper.find('#submit').trigger('click')
    expect(spy).toHaveBeenCalledWith({
      customer_name: data.customer_name,
      document_id: data.document_id
    })
  })

  const fillForm = async() => {
    const data = {
      customer_name: faker.random.word(),
      document_id: faker.random.number(),
      file: new File(['foo'], 'foo.png')
    }
    await wrapper.find('#customer-name').setValue(data.customer_name)
    await wrapper.find('#document-id').setValue(data.document_id)
    await wrapper.findComponent({ ref: 'files' }).vm.$emit('change', [data.file])

    return data
  }
})
