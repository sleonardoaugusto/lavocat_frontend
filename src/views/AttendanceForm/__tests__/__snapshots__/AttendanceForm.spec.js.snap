// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AttendanceForm /> Must render component correctly 1`] = `
<form novalidate="novalidate" class="v-form">
  <div class="row">
    <div class="col">
      <div class="text-h4 grey--text text--darken-3 font-weight-regular">Atendimento</div>
    </div>
  </div>
  <div class="row">
    <div class="col-sm-12 col-md-6 col-12">
      <div class="v-input theme--light v-text-field">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="customer-name" class="v-label theme--light" style="left: 0px; position: absolute;">Nome</label><input id="customer-name" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col-sm-12 col-md-6 col-12">
      <div class="v-input theme--light v-text-field">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="document-id" class="v-label theme--light" style="left: 0px; position: absolute;">CPF</label><input id="document-id" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col-sm-12 col-md-6 col-12">
      <div class="v-input theme--light v-text-field v-file-input">
        <div class="v-input__prepend-outer">
          <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light"></button></div>
        </div>
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="files" class="v-label theme--light" style="left: 0px; position: absolute;">Anexos</label>
              <div class="v-file-input__text v-file-input__text--chips"></div><input multiple="multiple" id="files" type="file">
            </div>
            <div class="v-input__append-inner">
              <div class="v-input__icon v-input__icon--clear"><button disabled="disabled" type="button" aria-label="clear icon" class="v-icon notranslate v-icon--disabled v-icon--link mdi mdi-close theme--light"></button></div>
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="row">
    <div class="col col-12">
      <div class="v-input v-textarea theme--light v-text-field">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="input-22" class="v-label theme--light" style="left: 0px; position: absolute;">Relat√≥rio</label><textarea id="input-22" rows="10"></textarea></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="row">
    <div class="text-right col"><button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg v-btn--right theme--light v-size--default success" id="submit"><span class="v-btn__content">
        Cadastrar
      </span></button></div>
  </div>
</form>
`;
